### PR TITLE
Teach `concat!()` to concatenate byte str literals

### LINKED ISSUE #52710: Concatenating bytestrings fails; rustc suggests non-existent `bconcat!()`
`concat!()` doesn't work on bytestrings (why?). When trying it, it fails with:
```
1 | const FOO: &[u8] = concat!("foo", "bar");
  |                    ^^^^^^^^^^^^^^^^^^^^^
  |                    |
  |                    expected slice, found str
  |                    help: consider adding a leading `b`: `bconcat!("foo", "bar")`
  |
  = note: expected type `&'static [u8]`
             found type `&'static str`
```

`bconcat!()` sounds interesting, but it isn't documented, and the compiler doesn't accept it:
```
error: cannot find macro `bconcat!` in this scope
 --> src/main.rs:1:20
  |
1 | const FOO: &[u8] = bconcat!("foo", "bar");
  |                    ^^^^^^^ help: you could try the macro: `concat`
```

I think the real fix here is that `bconcat!()` should exist.

### PR DESCRIPTION
**Pull request by estebank**
Fix #52710.

### CODE REVIEW

**Comment by estebank** [comment]
CC @rust-lang/lang

**Comment by kennytm** [comment]
I think the new behavior of `concat!` should be placed behind a feature gate.

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -1300,6 +1300,11 @@ impl LitKind {
         }
     }
 
+    /// Returns a `LitKind::ByteStr` from `content`.
```
Comment (eddyb) [inline src/libsyntax/ast.rs]: Why does this need to have a constructor method, specifically?

File: src/libsyntax_ext/concat.rs [hunk 2]
```diff
@@ -61,6 +73,36 @@ pub fn expand_syntax_ext(
...
+                .collect(),
+        );
+        err.emit();
+    }
     let sp = sp.apply_mark(cx.current_expansion.mark);
-    base::MacEager::expr(cx.expr_str(sp, Symbol::intern(&accumulator)))
+    if b_accumulator.len() > 0 {
+        base::MacEager::expr(cx.expr_lit(
+            sp,
+            ast::LitKind::new_byte_str(b_accumulator),
```
Comment (eddyb) [inline src/libsyntax_ext/concat.rs]: You might want instead a `cx.expr_byte_str`.

File: src/libsyntax/ast.rs (hunk 1 above)
Comment (estebank) [reply src/libsyntax/ast.rs]: removed

File: src/libsyntax_ext/concat.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax_ext/concat.rs]: done

**Comment by estebank** [comment]
> I think the new behavior of concat! should be placed behind a feature gate.

@kennytm other than bad style of introducing insta-stable behavior, do you foresee any problems with the new behavior? If the PR also included support to mix and match `""` and `b""` (which can be done, we just need to figure out what the appropriate default behavior should be and probably add a lint about it being supported but discouraged), I would definitely agree, but as it is I feel the change is small enough not to warrant a long release cycle.

**Comment by joshtriplett** [comment]
@estebank The usual concern there would be if there's some subtle aspect of the new behavior that we don't catch before shipping it, people start relying on that aspect, and we then can't change it even if we want to.

File: src/libsyntax_ext/concat.rs [hunk 3]
```diff
@@ -62,5 +83,32 @@ pub fn expand_syntax_ext(
...
+        }
+        err.help("do not mix byte string literals and string literals");
+        err.multipart_suggestion(
+            "you can use byte string literals",
+            string_pos
+                .iter()
+                .map(|pos| (pos.shrink_to_lo(), "b".to_string()))
+                .collect(),
+        );
+        err.emit();
```
Comment (estebank) [inline src/libsyntax_ext/concat.rs]: Just noticed that this error (and suggestion) will be wrong when doing `concat!(b"asdf", true, 3)`.

**Comment by estebank** [comment]
Oh boy... I really messed this up, trying to push to the RFCS repo. Refiled at #52932.

**Pull request closed without merging**